File: data-structures/linked-list/doubly-linked-list.ts
class DoublyListNode<T> {
  value: T;
  next: DoublyListNode<T> | null;
  prev: DoublyListNode<T> | null;

  constructor(
    value: T,
    next: DoublyListNode<T> | null = null,
    prev: DoublyListNode<T> | null = null
  ) {
    this.value = value;
    this.next = next;
    this.prev = prev;
  }
}
class DoublyLinkedList<T> {
  head: DoublyListNode<T> | null = null;
  tail: DoublyListNode<T> | null = null;
  length = 0;

  prepend(value: T): void {
    const node = new DoublyListNode(value);

    if (!this.head) {
      this.head = this.tail = node;
      this.length++;
      return;
    }

    node.next = this.head;
    this.head.prev = node;
    this.head = node;
    this.length++;
  }

  append(value: T): void {
    const node = new DoublyListNode(value);

    if (!this.head) {
      this.head = this.tail = node;
      this.length++;
      return;
    }

    this.tail!.next = node;
    node.prev = this.tail;
    this.tail = node;
    this.length++;
  }

  removeHead(): DoublyListNode<T> | null {
    if (!this.head) return null;

    const removed = this.head;

    if (this.length === 1) {
      this.head = this.tail = null;
      this.length = 0;
      return removed;
    }

    this.head = removed.next;
    this.head!.prev = null;
    this.length--;

    removed.next = null;
    removed.prev = null;
    return removed;
  }

  removeTail(): DoublyListNode<T> | null {
    if (!this.tail) return null;

    const removed = this.tail;

    if (this.length === 1) {
      this.head = this.tail = null;
      this.length = 0;
      return removed;
    }

    this.tail = removed.prev;
    this.tail!.next = null;
    this.length--;

    removed.prev = null;
    removed.next = null;
    return removed;
  }

  get(index: number): DoublyListNode<T> | null {
    if (index < 0 || index >= this.length) return null;

    // Optional optimization (DLL advantage):
    // if (index < this.length / 2) go from head else from tail
    let curr: DoublyListNode<T> | null;
    if (index < this.length / 2) {
      curr = this.head;
      while (index > 0) {
        curr = curr!.next;
        index--;
      }
      return curr;
    } else {
      curr = this.tail;
      let i = this.length - 1;
      while (i > index) {
        curr = curr!.prev;
        i--;
      }
      return curr;
    }
  }

  removeAt(index: number): DoublyListNode<T> | null {
    if (index < 0 || index >= this.length) return null;

    if (index === 0) return this.removeHead();
    if (index === this.length - 1) return this.removeTail();

    const node = this.get(index);
    if (!node) return null;

    const prev = node.prev!;
    const next = node.next!;

    prev.next = next;
    next.prev = prev;

    this.length--;

    node.prev = null;
    node.next = null;
    return node;
  }

  toArray(): T[] {
    const result: T[] = [];
    let curr = this.head;
    while (curr !== null) {
      result.push(curr.value);
      curr = curr.next;
    }
    return result;
  }

  reverse(): void {
    if (!this.head || this.length === 1) return;

    const oldHead = this.head;
    const oldTail = this.tail!;

    let cur: DoublyListNode<T> | null = this.head;
    while (cur !== null) {
      const next = cur.next;
      cur.next = cur.prev;
      cur.prev = next;
      cur = next;
    }

    this.head = oldTail;
    this.tail = oldHead;
  }
}

File: data-structures/linked-list/doubly_linked_list.py
from typing import Generic, TypeVar, Optional, List

T = TypeVar('T')


class DoublyListNode(Generic[T]):
    def __init__(
        self,
        value: T,
        next: Optional['DoublyListNode[T]'] = None,
        prev: Optional['DoublyListNode[T]'] = None
    ):
        self.value = value
        self.next = next
        self.prev = prev


class DoublyLinkedList(Generic[T]):
    def __init__(self):
        self.head: Optional[DoublyListNode[T]] = None
        self.tail: Optional[DoublyListNode[T]] = None
        self.length: int = 0

    def prepend(self, value: T) -> None:
        node = DoublyListNode(value)

        if not self.head:
            self.head = self.tail = node
            self.length += 1
            return

        node.next = self.head
        self.head.prev = node
        self.head = node
        self.length += 1

    def append(self, value: T) -> None:
        node = DoublyListNode(value)

        if not self.head:
            self.head = self.tail = node
            self.length += 1
            return

        self.tail.next = node
        node.prev = self.tail
        self.tail = node
        self.length += 1

    def remove_head(self) -> Optional[DoublyListNode[T]]:
        if not self.head:
            return None

        removed = self.head

        if self.length == 1:
            self.head = self.tail = None
            self.length = 0
            return removed

        self.head = removed.next
        self.head.prev = None
        self.length -= 1

        removed.next = None
        removed.prev = None
        return removed

    def remove_tail(self) -> Optional[DoublyListNode[T]]:
        if not self.tail:
            return None

        removed = self.tail

        if self.length == 1:
            self.head = self.tail = None
            self.length = 0
            return removed

        self.tail = removed.prev
        self.tail.next = None
        self.length -= 1

        removed.prev = None
        removed.next = None
        return removed

    def get(self, index: int) -> Optional[DoublyListNode[T]]:
        if index < 0 or index >= self.length:
            return None

        # Optimization (DLL advantage):
        # if index < length / 2 go from head else from tail
        if index < self.length // 2:
            curr = self.head
            while index > 0:
                curr = curr.next
                index -= 1
            return curr
        else:
            curr = self.tail
            i = self.length - 1
            while i > index:
                curr = curr.prev
                i -= 1
            return curr

    def remove_at(self, index: int) -> Optional[DoublyListNode[T]]:
        if index < 0 or index >= self.length:
            return None

        if index == 0:
            return self.remove_head()
        if index == self.length - 1:
            return self.remove_tail()

        node = self.get(index)
        if not node:
            return None

        prev_node = node.prev
        next_node = node.next

        prev_node.next = next_node
        next_node.prev = prev_node

        self.length -= 1

        node.prev = None
        node.next = None
        return node

    def to_array(self) -> List[T]:
        result: List[T] = []
        curr = self.head
        while curr is not None:
            result.append(curr.value)
            curr = curr.next
        return result

    def reverse(self) -> None:
        if not self.head or self.length == 1:
            return

        old_head = self.head
        old_tail = self.tail

        curr = self.head
        while curr is not None:
            next_node = curr.next
            curr.next = curr.prev
            curr.prev = next_node
            curr = next_node

        self.head = old_tail
        self.tail = old_head